client/src/main/scala/org/http4s/client/PoolManager.scala

```diff
@@ -84,30 +120,36 @@ private final class PoolManager[F[_], A <: Connection[F]](
         if (!isClosed) {
           @tailrec
           def go(): Unit =
-            idleQueue.dequeueFirst(_.requestKey == key) match {
+            getConnectionFromQueue(key) match {
               case Some(conn) if !conn.isClosed =>
                 logger.debug(s"Recycling connection: $stats")
                 callback(Right(NextConnection(conn, fresh = false)))
 
               case Some(closedConn) =>
                 logger.debug(s"Evicting closed connection: $stats")
-                allocated -= 1
+                decrConnection(key)
                 go()
 
-              case None if allocated < maxTotal =>
+              case None if numConnectionsCheckHolds(key) =>
                 logger.debug(s"Active connection not found. Creating new one. $stats")
                 createConnection(key, callback)
 
-              case None if idleQueue.nonEmpty =>
+              case None if curTotal == maxTotal =>
```

Comment:
I think we need to check that `maxConnectionsPerRequestKey(key)` didn't also fail since in that case it would break our invariant by making yet another one.

Reply:
New connections are only made in createConnection and we check for constraints there. How will be break the invariant?